### ISSUE TITLE
New Linter: uniquemarkers

### ISSUE THREAD

**Issue by JoelSpeed** [kubernetes-sigs/kube-api-linter#83]
There are a large number of markers in the kubebuilder space, that must be specified once, or, where they are specified multiple times, must have the same value (though there is no value in having them multiple times).

For example, the following list should all be specified once at most
```
+default
+kubebuilder:default
+kubebuilder:validation:Enum
+kubebuilder:validation:ExclusiveMaximum
+kubebuilder:validation:ExclusiveMinimum
+kubebuilder:validation:Format
+kubebuilder:validation:MaxItems
+kubebuilder:validation:MaxLength
+kubebuilder:validation:MaxProperties
+kubebuilder:validation:Maximum
+kubebuilder:validation:MinItems
+kubebuilder:validation:MinLength
+kubebuilder:validation:MinProperties
+kubebuilder:validation:Minimum
+kubebuilder:validation:Type
+kubebuilder:validation:UniqueItems
```

We should add a new, `uniquemarkers` linter that checks for these specific markers, and determines if any of them have multiple values, and, if they do have multiple values, that those values must all be identical, else to flag that there are multiple different values and that only one of them would apply.

This list should also be extended to include the K8s declarative validation markers once that project moves forward.

I think it would also be beneficial to allow configuration of additional unique markers, so that folks who have their own markers can add their own and benefit from this functionality.

**Comment by JoelSpeed** [comment]
@sivchari To respond to your https://github.com/kubernetes-sigs/kube-api-linter/pull/61#issuecomment-2863287777, I had imagined we would basically have a loop that looks something like
```
for _, m := range fieldMarkers {
  if mustBeUniqueMarkers.Has(m.Name) {
    // The particular marker must only have one value
    if len(m) == 1 {
      continue
    }

    // Check each of the different marker instances to see if they have differing "expressions"
  }
}
```

That way turning on one uniquelinters rule, would enable checking of many different markers

**Comment by everettraven** [comment]
@JoelSpeed That was my thinking as well. I think this issue is well scoped for linter check to be responsible for this. I think wherever possible we should avoid overloading other lint rules with more responsibilities than necessary.

I'm also happy to take this work on.

/assign

**Comment by sivchari** [comment]
I had a similar image of the basic policy of implementation. The difference seems to be whether to provide these separately or not. I was hoping to just activate one linter and have it report the full duplicate and everything that each marker would consider a duplicate.

**Comment by JoelSpeed** [comment]
So that would combine what's described here with the existing work for `duplicatemarkers`?

**Comment by sivchari** [comment]
Yes, I think it's good. Then the configuration like this.

```yaml
duplicatemarkers:
    rules:
      - default
      - maxitems
 ```

**Comment by JoelSpeed** [comment]
I think whatever we do, we create a list of default "must be unique" markers. And then expose additional config to allow that to be extended.

For example, perhaps I have a custom `+my:own:scheme:validation:Unique=true|false` that under the hood applies some schema change, and I want the linter to make sure there's only one value for this. I'd want to be able to configure

```
uniqueMarkers:
  - `my:own:scheme:validation:Unique`
```

**Comment by everettraven** [comment]
@sivchari I'm not sure I'm following what the expectation is when doing something like the expected configuration you are mentioning, could you elaborate what you are expecting to happen with a configuration like that?

@JoelSpeed I agree.

**Comment by sivchari** [comment]
I think both duplicatemarkers and uniquemarkers are both in the same context in the sense of eliminating exact matches or duplicates at a given marker. Therefore, rather than provide a separate linter for the two, I preferred to extend them as a configuration item within duplicatemarkers.
In the above example, it is assumed that when maxitems or default is set, an additional linter is called internally for each specific rule.
However, I did not consider user extensions. I thought about it again, but I thought the complexity of the code would be very large, so let's still go with the policy you both envisioned.
Sorry for the confusion.

**Comment by everettraven** [comment]
No worries.

Personally, I think it makes sense to have a separation of responsibilities for the linters and keeping them scoped to:
- `duplicatemarkers` - focuses on identifying exact duplicates. This is likely reasonable for _all_ markers because if it is identical to an already specified one, it likely does the exact same thing.
- `uniquemarkers` - focuses on identifying markers that must be unique - there is some potential overlap with `duplicatemarkers` if they are exact duplicates, but I think this check is a bit more nuanced and warrants being a separate check.

I generally like the idea of each linter being tightly scoped and following the unix philosophy of "do one thing and do it well".
